src/chart/utils/linkedList.utils.ts

```diff
@@ -0,0 +1,124 @@
+export class ListNode<T> {
+	public data: T;
+	public next: ListNode<T> | null;
+
+	constructor(data: T) {
+		this.data = data;
+		this.next = null;
+	}
+}
+/**
+ * Implementation of Linked list data structure.
+ * @param _head
+ * @param _tail
+ */
+export class LinkedList<T> {
+	private _head: ListNode<T> | null = null;
+	private _tail: ListNode<T> | null = null;
+	private length = 0;
+
+	constructor(head?: ListNode<T>, tail?: ListNode<T>) {
+		this._head = head ?? null;
+		this._tail = tail ?? null;
+	}
+
+	public insertAtEnd(data: T): ListNode<T> {
+		const node = new ListNode(data);
+		let current: ListNode<T>;
+
+		if (this.head === null) {
+			this._head = node;
+		} else {
+			current = this.head;
+			while (current.next) {
+				current = current.next;
+			}
+			current.next = node;
+			this._tail = node;
+		}
+		this.length++;
+		return node;
+	}
+
+	public insertAt(position: number, data: T) {
+		if (position > -1 && position < this.length && this.head) {
+			let current = this.head;
+			let index = 0;
+			let previous = current;
```

Comment:
> prev should be null on init, isn't it?

It's not necessary because the interaction with previous only happens when list has > 1 node and it reassigns later: 

previous = current;
current = current.next;

But logically yes, it's better to initialize as a null